{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "44a64e6c-37e0-4727-9c19-bf81e91d90e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO, Entrez\n",
    "import pandas as pd\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "92abba5a-f22b-4b91-9c88-2a30b7ed5681",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "'''\n",
    "Downloads entries from GenBank Nucleotide by given query\n",
    "Saves entries to file in GenBank format\n",
    "Input:\n",
    "    query - str - query for search in GenBank\n",
    "Output:\n",
    "    outfile - str - shortname of output file with sequences in genbank format\n",
    "'''\n",
    "def fetch_seq_from_Nucleotide(query, outfile):\n",
    "\n",
    "    Entrez.email = \"A.N.Other@example.com\"\n",
    "    # list with ids obtained by query\n",
    "    record = Entrez.read(Entrez.esearch(db=\"nucleotide\", term=query, idtype=\"acc\", RetMax=1000000))\n",
    "    id_list = record['IdList']\n",
    "    print(\"Query to GenBank Nucleotide database:\\\"{}\\\"\".format(query))\n",
    "    print(\"Number of records found: {}\".format(record[\"Count\"]))\n",
    "    print(\"{} records will be downloaded\".format(len(id_list)))\n",
    "    # output file with sequences will be saved in current working dir\n",
    "    f_name = os.path.abspath(outfile)\n",
    "\n",
    "    # output file with sequences in fasta format\n",
    "    file_out = open(f_name, 'w')\n",
    "    \n",
    "    # Saving sequences \n",
    "    \n",
    "    i = 0\n",
    "    # fetch each sequence from GenBank by its ID and writes to ouput file\n",
    "    for id in id_list:\n",
    "        i+=1\n",
    "        if i % 500 ==0:\n",
    "            print(\"Downloaded {} sequences\".format(i))\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=id, rettype=\"genbank\", retmode=\"text\")\n",
    "        for line in handle:\n",
    "            file_out.write(line)\n",
    "        handle.close()\n",
    "    file_out.close()\n",
    "    print('Finished')\n",
    "    return(f_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "a7bb4ed5-1832-4676-8515-a9339e4e1f5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_metadata_from_gb(input_file):\n",
    "    '''\n",
    "    For each entry in GenBank file retrieves the following data:\n",
    "\n",
    "    # Identifiers\n",
    "    Accession\n",
    "    GenBank title (DEFINITION)\n",
    "\n",
    "    # Classification\n",
    "    Organism name\n",
    "    Species\n",
    "    Isolate\n",
    "    Strain\n",
    "    Family\n",
    "    Lineage\n",
    "\n",
    "    #Sequence quality\n",
    "    Length\n",
    "    Strand\n",
    "\n",
    "    Num of annotated CDS\n",
    "\n",
    "    ORF1a\n",
    "    ORF1b\n",
    "    ORF2\n",
    "    coords\n",
    "    \n",
    "    #Sources\n",
    "    Country\n",
    "    Tissue/Specimen/Source\n",
    "    Collection date\n",
    "    Release date\n",
    "    Submitters\n",
    "    \n",
    "    '''\n",
    "\n",
    "    '''\n",
    "     dictionary\n",
    "     entries_data[GenBank Accession] = {}\n",
    "     entries_data[GenBank Accession].keys() = [\n",
    "                                                'Version',\n",
    "                                                'GenBank title'\n",
    "                                                'Organism name',\n",
    "                                                'Species',\n",
    "                                                'Family',\n",
    "                                                'Virus Lineage',\n",
    "                                                'Length',\n",
    "                                                'Isolate',\n",
    "                                                'Strain',\n",
    "                                                'Strand',\n",
    "                                                'Geo location'\n",
    "                                                'Country',\n",
    "                                                'Tissue/Specimen/Source',\n",
    "                                                'Host',\n",
    "                                                'Collection date',\n",
    "                                                'Release date',\n",
    "                                                'Submitters',\n",
    "                                                'Sequencing Technology'\n",
    "                                                'CDS number',\n",
    "                                                'Artificial',\n",
    "                                                'UNVERIFIED'\n",
    "                                                ]\n",
    "    '''\n",
    "    entries_data = {}\n",
    "\n",
    "    source_qualifiers = [\"strain\", \"isolate\", \"geo_loc_name\", \"country\", \"collection_date\", \"isolation_source\", \"host\"]\n",
    "    source_qualifiers_columns = [\"Strain\", \"Isolate\", \"Geo location\", \"Country\", \"Collection date\", \"Tissue/Specimen/Source\", \"Host\"]\n",
    "\n",
    "    entries = SeqIO.parse(os.path.abspath(input_file), \"genbank\")\n",
    "\n",
    "    for entry in entries:\n",
    "        entries_data[entry.name] = {}\n",
    "        entries_data[entry.name]['Version'] = entry.id\n",
    "        entries_data[entry.name]['GenBank title'] = entry.description\n",
    "\n",
    "        entries_data[entry.name]['Release date'] = entry.annotations['date']\n",
    "        entries_data[entry.name]['Organism name'] = entry.annotations['organism']\n",
    "        \n",
    "        entries_data[entry.name]['Virus Lineage'] = ';'.join(entry.annotations['taxonomy'])\n",
    "        num_taxa = len(entry.annotations['taxonomy'])\n",
    "        if num_taxa == 9:\n",
    "            entries_data[entry.name]['Species'] = entry.annotations['taxonomy'][-1]\n",
    "        else:\n",
    "             entries_data[entry.name]['Species'] = 'NA'\n",
    "        entries_data[entry.name]['Family'] = entry.annotations['taxonomy'][6]\n",
    "        \n",
    "        entries_data[entry.name]['Length'] = len(entry)\n",
    "\n",
    "        for reference in entry.annotations['references']:\n",
    "            if reference.title == 'Direct Submission':\n",
    "                entries_data[entry.name]['Submitters'] = reference.authors\n",
    "\n",
    "        count_cds = 0\n",
    "        for feature in entry.features:\n",
    "            if feature.type == 'source':\n",
    "                for qualif, colname in zip(source_qualifiers, source_qualifiers_columns):\n",
    "                    if qualif in feature.qualifiers:\n",
    "                        entries_data[entry.name][colname] = feature.qualifiers[qualif][0]\n",
    "                    else:\n",
    "                        entries_data[entry.name][colname] = 'NA'\n",
    "                if 'environmental_sample' in feature.qualifiers:\n",
    "                    entries_data[entry.name]['Environmental'] = 'Yes'\n",
    "                else:\n",
    "                    entries_data[entry.name]['Environmental'] = 'No'\n",
    "                    \n",
    "            if feature.type == 'CDS':\n",
    "                \n",
    "                count_cds +=1\n",
    "                \n",
    "        entries_data[entry.name]['CDS count'] = count_cds\n",
    "\n",
    "\n",
    "        artificial_condition1 = (\"patent\" in entry.description.lower()) or (\"oligonucleotide\" in entry.description.lower())\n",
    "        artificial_condition2 = \"crispr\" in entry.description.lower()\n",
    "        artificial_condition3 = \"method\" in entry.description.lower()\n",
    "        artificial_condition4 = (\"modified microbial nucleic acid\" in entry.description.lower()) or (\"nonfunctional\" in entry.description.lower())\n",
    "\n",
    "        \n",
    "        \n",
    "        if artificial_condition1 or artificial_condition2 or artificial_condition3 or artificial_condition4:\n",
    "            entries_data[entry.name]['Artificial'] = \"Yes\"\n",
    "        else:\n",
    "            entries_data[entry.name]['Artificial'] = \"No\"\n",
    "\n",
    "\n",
    "        if 'UNVERIFIED' in entry.description:\n",
    "            entries_data[entry.name]['UNVERIFIED'] = \"Yes\"\n",
    "        else:\n",
    "            entries_data[entry.name]['UNVERIFIED'] = \"No\"\n",
    "\n",
    "    meta_dataframe = pd.DataFrame.from_dict(entries_data, orient='index')\n",
    "    \n",
    "    meta_dataframe.to_csv(os.path.splitext(os.path.abspath(input_file))[0] + '.csv')\n",
    "    #country_map - file with abbreviations of countries\n",
    "    return meta_dataframe\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "87cfd169-5ef8-4943-9047-98f218a3b9fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "#custom csv parser \n",
    "def read_csv(file_name, strip_it=True):\n",
    "    if not os.path.exists(file_name):\n",
    "        return {}\n",
    "\n",
    "    def strip(value):\n",
    "        return strip_it and value.strip() or value\n",
    "\n",
    "    with open(file_name) as csvfile:\n",
    "        reader = csv.DictReader(csvfile,\n",
    "                                delimiter=\",\", \n",
    "                                fieldnames=[\"base\", \"new\"])\n",
    "        result = {}\n",
    "        for row in reader:\n",
    "\n",
    "            result[strip(row[\"base\"])] = strip(row[\"new\"])\n",
    "        csvfile.close()\n",
    "        return result\n",
    "\n",
    "def map_feature(feature, feature_map):\n",
    "    '''\n",
    "    feature_map - dictionary, e.g. feature_map['Italia']='ITA'\n",
    "    feature - possible key from feature_map\n",
    "    return value if key is in feature_map\n",
    "    '''\n",
    "    for k, v in feature_map.items():\n",
    "        if feature.lower() == k.lower():\n",
    "            return v\n",
    "    return feature\n",
    "\n",
    "\n",
    "def map_keywords(stri, keywords_dict):\n",
    "    '''\n",
    "    stri - GenBank DEFINITION, string\n",
    "    keywords_dict - dictionary with keywords that can indicate which ORF is covered  by the sequence\n",
    "    return list of unique ORF codes if the corresponding keywords werencountered in DEFINITION\n",
    "    '''\n",
    "    codes = []\n",
    "    for word, code in keywords_dict.items():\n",
    "        if word.lower() in stri.lower():\n",
    "            codes.append(code)\n",
    "    codes = list(set(codes))\n",
    "    return codes\n",
    "\n",
    "    \n",
    "def orf_coord(input_file, orf_map):\n",
    "    '''\n",
    "    Retrieves the coordinates of ORFs\n",
    "    \n",
    "    Input:\n",
    "        input_file - file with nucleotide sequences in genbank-format\n",
    "        orf_map - csv file annotation of orfs and their codes\n",
    "    Output:\n",
    "        coord_file - file with coordinates\n",
    "    '''\n",
    "\n",
    "    # dictionary with possiple names of astrovirus ORFs indicated in note, gene and product qualifiers\n",
    "    # orf_dict[\"ORF1a\"] = \"1A\"\n",
    "    orf_dict = read_csv(orf_map)\n",
    "\n",
    "    # ORFs which coordinates this script will extract\n",
    "    orf_types_final = ['1A', '1B', '2']\n",
    "    # ORFs that can be met\n",
    "    orf_types = ['1A', '1B', '1AB', '2']\n",
    "\n",
    "    # name of output file with ORF coordinates\n",
    "    out_file_name_temp = os.path.splitext(os.path.abspath(input_file))[0]\n",
    "    out_file_name = out_file_name_temp  + '_orf-coords.csv'\n",
    "    print(out_file_name)\n",
    "\n",
    "    # Keywords in description that can indicate which ORF is covered  by the sequence\n",
    "    description_keywords = { \"ORF2-like\": \"2\", \n",
    "                            \"ORF2\": \"2\",\n",
    "                            \"precapsid\":\"2\", \n",
    "                            \"CP sequence\": \"2\", \n",
    "                            \"capsid\": \"2\",\n",
    "                            \"ORF1a-like\": \"1A\", \n",
    "                            \"nsp1a-like\": \"1A\", \n",
    "                            \"ORF1A\": \"1A\",\n",
    "                            \"1b-like\": \"1B\", \n",
    "                            \"RNA polymerase\": \"1B\", \n",
    "                            \"RdRp\": \"1B\",\n",
    "                            \"ORF1b\": \"1B\"\n",
    "                            }\n",
    "    \n",
    "    # ENTRIES with known errors in annotation\n",
    "    exceptions_ORF2 = [\"MN725025\", \"KY047739\", \"S68561\", \"KP404151\",\"KP404152\"] # the last CDS in conflict ORF1b ORF2 is ORF2\n",
    "    # PV999257 is actually ORF1a\n",
    "    exceptions_1a = [\"PV999257\"]\n",
    "    #AB518702 AB518703 is actually RdRp\n",
    "    exceptions_1b = []\n",
    "    \n",
    "    # dictionary with ORF coordinates\n",
    "    dict_coord = {}\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "    with open(input_file) as handle:\n",
    "        records = list(SeqIO.parse(handle, 'gb'))\n",
    "\n",
    "        # entries with no annotations\n",
    "        no_annot = 0\n",
    "        # entries with no annotations of target CDS\n",
    "        no_annot_target = 0\n",
    "        # CDS with no annotations of target ORFs\n",
    "        notarget_annot_CDS = 0\n",
    "        # CDS with no annotations\n",
    "        no_annot_CDS = 0\n",
    "\n",
    "        # Entries with CDS that has annotation \"nonstructural protein\"\n",
    "        entries_onlynonstrprotein = []\n",
    "        entries_onlynonstrprotein_short = []\n",
    "        entries_conflicting_annotations = []\n",
    "        entries_noannot = []\n",
    "        entries_notarget_annot = []\n",
    "        \n",
    "        for rec in records:\n",
    "\n",
    "            # Omit unverified sequences\n",
    "            if \"UNVERIFIED\" in rec.description:\n",
    "                continue\n",
    "            # The coordinates of artificial sequences will not be extracted\n",
    "            artificial_condition1 = (\"patent\" in rec.description.lower())\n",
    "            artificial_condition2 = (\"crispr\" in rec.description.lower()) or (\"oligonucleotide\" in rec.description.lower())\n",
    "            artificial_condition3 = \"method\" in rec.description.lower()\n",
    "            artificial_condition4 = \"modified microbial nucleic acid\" in rec.description.lower()\n",
    "            # Omit artificial sequences\n",
    "            if artificial_condition1 or artificial_condition2 or artificial_condition3 or artificial_condition4:\n",
    "                continue\n",
    "\n",
    "            \n",
    "            dict_coord[rec.name] = {}\n",
    "            # iterate over record features\n",
    "\n",
    "            # number of CDS\n",
    "            cds_count = 0\n",
    "            # number of annotated CDS with target ORFs\n",
    "            cds_annot_count = 0\n",
    "\n",
    "            # Set coordinates for nonstructural protein which type cannot be inferred from \n",
    "            NSP_start, NSP_end, NSP_strand = None, None, None\n",
    "            for feature in rec.features:\n",
    "\n",
    "                # check primers to identify the amplified region\n",
    "                # !!! REVISE removing Mon269_Mon270\n",
    "                if feature.type == 'source':\n",
    "                    Mon340_Mon348 = False\n",
    "                    Mon269_Mon270 = False\n",
    "                    if feature.qualifiers.get(\"PCR_primers\") != None:\n",
    "                        #print(feature.qualifiers[\"PCR_primers\"])\n",
    "                        if \"mon340\" in feature.qualifiers[\"PCR_primers\"][0].lower() or \"mon348\" in feature.qualifiers[\"PCR_primers\"][0].lower():\n",
    "                            Mon340_Mon348 = True\n",
    "                            Mon269_Mon270 = False\n",
    "                        elif \"mon269\" in feature.qualifiers[\"PCR_primers\"][0].lower() or \"mon270\" in feature.qualifiers[\"PCR_primers\"][0].lower():\n",
    "                            Mon340_Mon348 = False\n",
    "                            Mon269_Mon270 = True\n",
    "        \n",
    "                if feature.type == 'CDS':\n",
    "                    cds_count += 1\n",
    "                    \n",
    "                    # Check codon start\n",
    "                    if 'codon_start' in feature.qualifiers.keys():\n",
    "                        cod_start = int(feature.qualifiers['codon_start'][0]) - 1\n",
    "                        if cod_start<0:\n",
    "                            print('Codon start < 0, something is wrong')\n",
    "                    else:\n",
    "                        cod_start = 0\n",
    "\n",
    "                    # check all annotations in gene, product, note qualifiers\n",
    "                    CDS_raw_annotations = []\n",
    "                    for el in ('product', 'gene', 'note'):\n",
    "                        CDS_raw_annotations.extend(feature.qualifiers.get(el, []))\n",
    "                    \n",
    "                    if len(CDS_raw_annotations) != 0:\n",
    "                        # now we will leave annotations of target ORFs\n",
    "                        CDS_annotations = list(map(lambda x: map_feature(x, orf_dict), CDS_raw_annotations))\n",
    "                        CDS_annotations = list(filter(lambda x: x in orf_types, CDS_annotations))\n",
    "\n",
    "                        # TARGET ORF ANNOTATIONS WERE  FOUND\n",
    "                        if len(CDS_annotations) != 0:\n",
    "                                                    \n",
    "                            #print(\"Annotations were successfully found\")\n",
    "                            #print(CDS_annotations)\n",
    "\n",
    "                            # check for consistency of CDS annotations\n",
    "                            uniq_annotations = list(set(CDS_annotations))\n",
    "                            if len(uniq_annotations) == 1:\n",
    "                                CDS_product = uniq_annotations[0]\n",
    "                                # CDS is a target ORF, the coordinates are not joined\n",
    "                                if CDS_product in orf_types_final:\n",
    "                                    dict_coord[rec.name][CDS_product] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                    dict_coord[rec.name][CDS_product + '-strand'] = feature.location.strand\n",
    "                                    cds_annot_count +=1\n",
    "                                elif CDS_product == \"1AB\":\n",
    "\n",
    "                                    # check whether the coordinates are joined\n",
    "                                    if len(feature.location.parts) == 2:\n",
    "                                        dict_coord[rec.name]['1A'] = [int(feature.location.parts[0].start) + cod_start, int(feature.location.parts[0].end)]\n",
    "                                        dict_coord[rec.name]['1A-strand'] = feature.location.parts[0].strand\n",
    "                                        dict_coord[rec.name]['1B'] = [int(feature.location.parts[1].start), int(feature.location.parts[1].end)]\n",
    "                                        dict_coord[rec.name]['1B-strand'] = feature.location.parts[1].strand\n",
    "                                        cds_annot_count +=1\n",
    "                                    elif len(feature.location.parts) == 1:\n",
    "                                        dict_coord[rec.name]['1B'] = [int(feature.location.parts[0].start) + cod_start, int(feature.location.parts[0].end)]\n",
    "                                        dict_coord[rec.name]['1B-strand'] = feature.location.parts[0].strand\n",
    "                                        cds_annot_count +=1\n",
    "                                    else:\n",
    "                                        print(\"strange locations in {}\".format(rec.name))\n",
    "                                        print(feature.location)\n",
    "                            else:\n",
    "                                print(\"Conflicting CDS annotations for {}\".format(rec.name))\n",
    "                                #print(CDS_annotations)\n",
    "                                #print(uniq_annotations)\n",
    "\n",
    "                                present_1A = ('1A' in uniq_annotations)\n",
    "                                present_1B = ('1B' in uniq_annotations)\n",
    "                                present_1A_1B = present_1A and present_1B\n",
    "                                present_1AB = ('1AB' in uniq_annotations)\n",
    "                                present_1A_1AB = present_1A and present_1AB\n",
    "                                present_1B_1AB = present_1B and present_1AB\n",
    "                                present_2 = ('2' in uniq_annotations)\n",
    "                                present_1AB_2 = (present_1A and present_2) or (present_1B and present_2) or (present_1AB and present_2) \n",
    "                                \n",
    "                                \n",
    "                                if present_1AB_2:\n",
    "                                    if rec.name in exceptions_ORF2:\n",
    "                                        dict_coord[rec.name]['2'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                        dict_coord[rec.name]['2-strand'] = feature.location.strand\n",
    "                                        cds_annot_count +=1\n",
    "                                        print(\"Resolved for ORF2\")\n",
    "                                    else:\n",
    "                                        print(\"Could not resolve the conflict\")\n",
    "                                        entries_conflicting_annotations.append(rec.name + ': ORF1ab + ORF2' + ':' + rec.description + '\\n')\n",
    "                                # If ORF1a and ORF1b are present, we choose ORF1b\n",
    "                                elif present_1A_1B:\n",
    "                                    if rec.name in exceptions_1a:\n",
    "                                        dict_coord[rec.name]['1A'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                        dict_coord[rec.name]['1A-strand'] = feature.location.strand\n",
    "                                        cds_annot_count +=1\n",
    "                                        print(\"Resolved for ORF1a\")\n",
    "                                    else:\n",
    "                                        dict_coord[rec.name]['1B'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                        dict_coord[rec.name]['1B-strand'] = feature.location.strand\n",
    "                                        cds_annot_count +=1\n",
    "                                        #entries_conflicting_annotations.append(rec.name + ': ORF1a + ORF1b' + ':' + rec.description + '\\n')\n",
    "                                        print(\"Resolved for ORF1b\")\n",
    "                                # ORF1ab with joined coordinates\n",
    "                                elif len(feature.location.parts) == 2:\n",
    "                                    dict_coord[rec.name]['1A'] = [int(feature.location.parts[0].start) + cod_start, int(feature.location.parts[0].end)]\n",
    "                                    dict_coord[rec.name]['1A-strand'] = feature.location.parts[0].strand\n",
    "                                    dict_coord[rec.name]['1B'] = [int(feature.location.parts[1].start), int(feature.location.parts[1].end)]\n",
    "                                    dict_coord[rec.name]['1B-strand'] = feature.location.parts[1].strand\n",
    "                                    cds_annot_count +=1\n",
    "                                    print(\"Resolved for ORF1ab\")\n",
    "                                elif present_1A_1AB and ('1A' not in dict_coord[rec.name].keys()):\n",
    "                                    dict_coord[rec.name]['1A'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                    dict_coord[rec.name]['1A-strand'] = feature.location.strand\n",
    "                                    cds_annot_count +=1\n",
    "                                    print(\"Resolved for ORF1a\")\n",
    "                                elif present_1B_1AB and ('1B' not in dict_coord[rec.name].keys()):\n",
    "                                    dict_coord[rec.name]['1B'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                    dict_coord[rec.name]['1B-strand'] = feature.location.strand\n",
    "                                    cds_annot_count +=1\n",
    "                                    print(\"Resolved for ORF1b\")\n",
    "                                elif '1B' in dict_coord[rec.name].keys() or '1A' in dict_coord[rec.name].keys():\n",
    "                                    print(\"The coordinates've been extracted from another CDS\")\n",
    "                            if rec.name == \"MF973517\":\n",
    "                                print(\"MF973517\")\n",
    "                                print(cds_annot_count)\n",
    "                                print(dict_coord[rec.name])\n",
    "                        # NO TARGET ORF ANNOTATIONS WERE FOUND\n",
    "                        else:            \n",
    "                            #print(\"No annotations of target ORFs of CDS in {}\".format(rec.name))\n",
    "                            #print(CDS_raw_annotations)\n",
    "                            #print(rec.description)\n",
    "                            notarget_annot_CDS +=1\n",
    "\n",
    "                            if \"nonstructural protein\" in CDS_raw_annotations:\n",
    "                                #print(\"CDS of {} has the only annotation \\'nonstructural protein\\' \".format(rec.name))\n",
    "                                if len(rec) > 300:\n",
    "                                    if \"orf1a\" in rec.description.lower():\n",
    "                                        dict_coord[rec.name][\"1A\"] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                        dict_coord[rec.name]['1A-strand'] = feature.location.strand\n",
    "                                        cds_annot_count +=1\n",
    "                                        cds_count +=1\n",
    "                                    # sequences covers ORF1b and ORF2\n",
    "                                    elif  len(rec)< 700 and (\"capsid\" in rec.description.lower()):\n",
    "                                        dict_coord[rec.name][\"1B\"] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                        dict_coord[rec.name]['1B-strand'] = feature.location.strand\n",
    "                                        cds_annot_count +=1\n",
    "                                        cds_count +=1\n",
    "                                    else:\n",
    "                                        # ORF1b with joined coordinates\n",
    "                                        #print(feature.location.parts)\n",
    "                                        if len(feature.location.parts) == 2:\n",
    "                                            dict_coord[rec.name]['1A'] = [int(feature.location.parts[0].start) + cod_start, int(feature.location.parts[0].end)]\n",
    "                                            dict_coord[rec.name]['1A-strand'] = feature.location.parts[0].strand\n",
    "                                            dict_coord[rec.name]['1B'] = [int(feature.location.parts[1].start), int(feature.location.parts[1].end)]\n",
    "                                            dict_coord[rec.name]['1B-strand'] = feature.location.parts[1].strand\n",
    "                                            cds_annot_count +=1\n",
    "                                        else:\n",
    "                                            NSP_start = int(feature.location.start) + cod_start\n",
    "                                            NSP_end = int(feature.location.parts[0].end)\n",
    "                                            NSP_strand = feature.location.strand\n",
    "                                            #entries_onlynonstrprotein.append(rec.name + ':' + str(len(rec)) + \":\"  +rec.description + '\\n')\n",
    "                                            # Тут записать координаты рамки и в конце проверить, нет ли других аннотаций\n",
    "\n",
    "                                else:\n",
    "                                    if Mon340_Mon348 == True:\n",
    "                                        dict_coord[rec.name][\"1A\"] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                        dict_coord[rec.name]['1A-strand'] = feature.location.strand\n",
    "                                        cds_annot_count +=1\n",
    "                                        cds_count +=1\n",
    "                                    else:\n",
    "                                        '''\n",
    "                                        These sequences cover ORF1a\n",
    "                                        KX522573\n",
    "                                        KX522572\n",
    "                                        KX522571\n",
    "                                        KX522570\n",
    "                                        KX522569\n",
    "                                        KX522568\n",
    "                                        KX522567\n",
    "                                        KX522566\n",
    "                                        KP090043\n",
    "                                        KP090042\n",
    "                                        KP090041\n",
    "                                        KP090040\n",
    "                                        KP090039\n",
    "                                        KP090038\n",
    "                                        KP090037\n",
    "                                        KP090036\n",
    "                                        KP090035\n",
    "                                        '''\n",
    "                                        if \"KX5225\" in rec.name or \"KP0900\" in rec.name:\n",
    "                                            dict_coord[rec.name][\"1A\"] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                            dict_coord[rec.name]['1A-strand'] = feature.location.strand\n",
    "                                            cds_annot_count +=1\n",
    "                                            cds_count +=1\n",
    "                                        else:\n",
    "                                            entries_onlynonstrprotein_short.append(rec.name + ':' + rec.description + '\\n')\n",
    "                    else:\n",
    "                        #print(\"No annotations of CDS in {}\".format(rec.name))\n",
    "                        #print(rec.description)\n",
    "                        #print(feature)\n",
    "                        no_annot_CDS +=1\n",
    "\n",
    "            # Determining the type of ORF indicated as \"nonstructural protein\"\n",
    "            if NSP_start != None:\n",
    "                present_1A = ('1A' in dict_coord[rec.name].keys())\n",
    "                present_1B = ('1B' in dict_coord[rec.name].keys())\n",
    "                present_2 = ('2' in dict_coord[rec.name].keys())\n",
    "                present_1B2 = present_1B and present_2\n",
    "                present_1A2 = present_1A and present_2\n",
    "\n",
    "                if present_1B2:\n",
    "                    dict_coord[rec.name][\"1A\"] = [NSP_start, NSP_end]\n",
    "                    dict_coord[rec.name]['1A-strand'] = NSP_strand\n",
    "                    cds_annot_count +=1\n",
    "                    cds_count +=1\n",
    "                elif present_1A2:\n",
    "                    dict_coord[rec.name][\"1B\"] = [NSP_start, NSP_end]\n",
    "                    dict_coord[rec.name]['1B-strand'] = NSP_strand\n",
    "                    cds_annot_count +=1\n",
    "                    cds_count +=1\n",
    "                elif present_1A:\n",
    "                    dict_coord[rec.name][\"1B\"] = [NSP_start, NSP_end]\n",
    "                    dict_coord[rec.name]['1B-strand'] = NSP_strand\n",
    "                    cds_annot_count +=1\n",
    "                    cds_count +=1\n",
    "                elif present_2:\n",
    "                    dict_coord[rec.name][\"1B\"] = [NSP_start, NSP_end]\n",
    "                    dict_coord[rec.name]['1B-strand'] = NSP_strand\n",
    "                    cds_annot_count +=1\n",
    "                    cds_count +=1\n",
    "                elif present_1B:\n",
    "                    dict_coord[rec.name][\"1A\"] = [NSP_start, NSP_end]\n",
    "                    dict_coord[rec.name]['1A-strand'] = NSP_strand\n",
    "                    cds_annot_count +=1\n",
    "                    cds_count +=1\n",
    "                else:\n",
    "                    entries_onlynonstrprotein.append(rec.name + ':' + str(len(rec)) + \":\"  +rec.description + '\\n')\n",
    "\n",
    "            \n",
    "            '''\n",
    "            # Check whether seq encodes ORF2\n",
    "            if Mon269_Mon270 == True:\n",
    "                print(\"Mon ORF2 primers: {}\".format(rec.name))\n",
    "                #dict_coord[rec.name]['2'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                #dict_coord[rec.name]['2-strand'] = feature.location.strand\n",
    "                #cds_annot_count +=1    \n",
    "            '''\n",
    "\n",
    "\n",
    "            if cds_annot_count == 0 or cds_count == 0:\n",
    "                # Check the keywords in DEFINITION\n",
    "                if len(rec) < 2700:\n",
    "                    ORFs = map_keywords(rec.description,description_keywords)\n",
    "                    if len(ORFs) == 1:\n",
    "                        ORF_code = ORFs[0]\n",
    "                        for feature in rec.features:\n",
    "                            if feature.type == 'misc_feature':\n",
    "                                dict_coord[rec.name][ORF_code] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                dict_coord[rec.name][ORF_code + '-strand'] = feature.location.strand\n",
    "                                cds_annot_count +=1\n",
    "                                cds_count +=1\n",
    "                    elif len(ORFs) == 0:\n",
    "                        print(\"No keywords were found in description for {}\".format(rec.name))\n",
    "                    else:\n",
    "                        print(\"Conflicting keywords were found in description for {}\".format(rec.name))\n",
    "                        print(ORFs)\n",
    "                else:\n",
    "                    # Check misc features for annotations\n",
    "                    for feature in rec.features:\n",
    "                        if feature.type == \"misc_feature\":\n",
    "                            miscfeature_raw_annotations = []\n",
    "                            for el in ('product', 'gene', 'note'):\n",
    "                                miscfeature_raw_annotations.extend(feature.qualifiers.get(el, []))\n",
    "                            #print(miscfeature_raw_annotations)\n",
    "                            \n",
    "                            miscfeature_annotations = []\n",
    "                            for f in miscfeature_raw_annotations:\n",
    "                                ORFs = map_keywords(f,description_keywords)\n",
    "                                miscfeature_annotations.extend(ORFs)\n",
    "                            miscfeature_annotations = list(set(miscfeature_annotations))\n",
    "                            if len(miscfeature_annotations) != 1:\n",
    "                                print(\"Conflicting annotations in misc_feature for {}\".format(rec.name))\n",
    "                                print(miscfeature_raw_annotations)\n",
    "                                print(miscfeature_annotations)\n",
    "                            else:\n",
    "                                ORF_code = miscfeature_annotations[0]\n",
    "                                dict_coord[rec.name][ORF_code] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                dict_coord[rec.name][ORF_code + '-strand'] = feature.location.strand\n",
    "                                cds_annot_count +=1\n",
    "                                cds_count +=1\n",
    "\n",
    "            if cds_annot_count == 0:\n",
    "                no_annot_target +=1\n",
    "                entries_notarget_annot.append(rec.name + ':' + str(len(rec)) + ':' + rec.description + '\\n')\n",
    "            if cds_count == 0:\n",
    "                no_annot +=1\n",
    "                entries_noannot.append(rec.name + ':' + str(len(rec)) + ':' + rec.description + '\\n')\n",
    "           \n",
    "\n",
    "\n",
    "\n",
    "        # Write extracted coordinates to file\n",
    "        with open(out_file_name, 'w') as out_file:\n",
    "            line = \"GBAC\"\n",
    "            for ORF in orf_types_final:\n",
    "                line = line + ',' + ORF + ',' + ORF + '-strand'\n",
    "            out_file.write(line + '\\n')\n",
    "            for rec_id, values in dict_coord.items():\n",
    "                s = rec_id\n",
    "                for orf in orf_types_final:\n",
    "                    if orf in values:\n",
    "                        s += f\",{values[orf][0]}-{values[orf][1]},{values[orf+'-strand']}\"\n",
    "                    else:\n",
    "                        s += \",NA-NA,NA\"\n",
    "                s += \"\\n\"\n",
    "                out_file.write(s)\n",
    "        out_file.close()\n",
    "\n",
    "\n",
    "        print(\"Entries with no CDS annotation: {}\".format(no_annot))\n",
    "        print(\"Entries with no CDS with target ORF annotation: {}\".format(no_annot_target))\n",
    "        print(\"CDS with no target ORFs: {}\".format(notarget_annot_CDS))\n",
    "        print(\"CDS with no annotations: {}\".format(no_annot_CDS))\n",
    "\n",
    "        with open(out_file_name_temp + '_noannotCDS.txt', 'w') as file:\n",
    "            file.writelines(entries_noannot)\n",
    "\n",
    "        with open(out_file_name_temp + '_noannot_targetCDS.txt', 'w') as file:\n",
    "            file.writelines(entries_notarget_annot)\n",
    "            \n",
    "        with open(out_file_name_temp + '_noannot_targetCDS.txt', 'w') as file:\n",
    "            file.writelines(entries_notarget_annot)\n",
    "\n",
    "        with open(out_file_name_temp + '_conflictingannot.txt', 'w') as file:\n",
    "            file.writelines(entries_conflicting_annotations)\n",
    "\n",
    "        with open(out_file_name_temp + '_onlyNSP_long.txt', 'w') as file:\n",
    "            file.writelines(entries_onlynonstrprotein)\n",
    "            \n",
    "        with open(out_file_name_temp + '_onlyNSP_short.txt', 'w') as file:\n",
    "            file.writelines(entries_onlynonstrprotein_short)\n",
    "\n",
    "        \n",
    "        return 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4bf1e178-edbb-46b3-99db-0a4d954003ee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query to GenBank Nucleotide database:\"\"txid39733\"[Organism]\"\n",
      "Number of records found: 16012\n",
      "16012 records will be downloaded\n",
      "Downloaded 500 sequences\n",
      "Downloaded 1000 sequences\n",
      "Downloaded 1500 sequences\n",
      "Downloaded 2000 sequences\n",
      "Downloaded 2500 sequences\n",
      "Downloaded 3000 sequences\n",
      "Downloaded 3500 sequences\n",
      "Downloaded 4000 sequences\n",
      "Downloaded 4500 sequences\n",
      "Downloaded 5000 sequences\n",
      "Downloaded 5500 sequences\n",
      "Downloaded 6000 sequences\n",
      "Downloaded 6500 sequences\n",
      "Downloaded 7000 sequences\n",
      "Downloaded 7500 sequences\n",
      "Downloaded 8000 sequences\n",
      "Downloaded 8500 sequences\n",
      "Downloaded 9000 sequences\n",
      "Downloaded 9500 sequences\n",
      "Downloaded 10000 sequences\n",
      "Downloaded 10500 sequences\n",
      "Downloaded 11000 sequences\n",
      "Downloaded 11500 sequences\n",
      "Downloaded 12000 sequences\n",
      "Downloaded 12500 sequences\n",
      "Downloaded 13000 sequences\n",
      "Downloaded 13500 sequences\n",
      "Downloaded 14000 sequences\n",
      "Downloaded 14500 sequences\n",
      "Downloaded 15000 sequences\n",
      "Downloaded 15500 sequences\n",
      "Downloaded 16000 sequences\n",
      "Finished\n"
     ]
    }
   ],
   "source": [
    "file_gb = fetch_seq_from_Nucleotide('\"txid39733\"[Organism]' , \"data/Astroviridae_15102025.gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "e8272523-aba2-49da-a300-12cb03eacb19",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = \"data/Astroviridae_15102025.gb\"\n",
    "data = fetch_metadata_from_gb(input_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e0b0c906-14ec-4e31-94c0-f6b600922733",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Version</th>\n",
       "      <th>GenBank title</th>\n",
       "      <th>Release date</th>\n",
       "      <th>Organism name</th>\n",
       "      <th>Virus Lineage</th>\n",
       "      <th>Species</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Submitters</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Isolate</th>\n",
       "      <th>Geo location</th>\n",
       "      <th>Country</th>\n",
       "      <th>Collection date</th>\n",
       "      <th>Tissue/Specimen/Source</th>\n",
       "      <th>Host</th>\n",
       "      <th>Environmental</th>\n",
       "      <th>CDS count</th>\n",
       "      <th>Artificial</th>\n",
       "      <th>UNVERIFIED</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PG036827</th>\n",
       "      <td>PG036827.1</td>\n",
       "      <td>JP 2022536474-A/8: Methods and Compositions of...</td>\n",
       "      <td>28-FEB-2025</td>\n",
       "      <td>Human astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>5388</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PG036826</th>\n",
       "      <td>PG036826.1</td>\n",
       "      <td>JP 2022536474-A/7: Methods and Compositions of...</td>\n",
       "      <td>28-FEB-2025</td>\n",
       "      <td>Human astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>5412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PG036825</th>\n",
       "      <td>PG036825.1</td>\n",
       "      <td>JP 2022536474-A/6: Methods and Compositions of...</td>\n",
       "      <td>28-FEB-2025</td>\n",
       "      <td>Human astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>5358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PG036824</th>\n",
       "      <td>PG036824.1</td>\n",
       "      <td>JP 2022536474-A/5: Methods and Compositions of...</td>\n",
       "      <td>28-FEB-2025</td>\n",
       "      <td>Human astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>5358</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PG036823</th>\n",
       "      <td>PG036823.1</td>\n",
       "      <td>JP 2022536474-A/4: Methods and Compositions of...</td>\n",
       "      <td>28-FEB-2025</td>\n",
       "      <td>Human astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>5337</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GN354831</th>\n",
       "      <td>GN354831.1</td>\n",
       "      <td>Sequence 4595 from Patent WO2007130519</td>\n",
       "      <td>12-MAY-2009</td>\n",
       "      <td>Mamastrovirus 13</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>Mamastrovirus ovis</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GN354830</th>\n",
       "      <td>GN354830.1</td>\n",
       "      <td>Sequence 4594 from Patent WO2007130519</td>\n",
       "      <td>12-MAY-2009</td>\n",
       "      <td>Mamastrovirus 13</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>Mamastrovirus ovis</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GN354829</th>\n",
       "      <td>GN354829.1</td>\n",
       "      <td>Sequence 4593 from Patent WO2007130519</td>\n",
       "      <td>12-MAY-2009</td>\n",
       "      <td>Mamastrovirus 13</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>Mamastrovirus ovis</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DL470173</th>\n",
       "      <td>DL470173.1</td>\n",
       "      <td>ANTISENSE ANTIVIRAL COMPOUND AND METHOD FOR TR...</td>\n",
       "      <td>22-DEC-2008</td>\n",
       "      <td>Human astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AX348182</th>\n",
       "      <td>AX348182.1</td>\n",
       "      <td>Sequence 52 from Patent WO0202811</td>\n",
       "      <td>06-FEB-2002</td>\n",
       "      <td>Human astrovirus 1</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>Mamastrovirus hominis</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>67</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Version                                      GenBank title  \\\n",
       "PG036827  PG036827.1  JP 2022536474-A/8: Methods and Compositions of...   \n",
       "PG036826  PG036826.1  JP 2022536474-A/7: Methods and Compositions of...   \n",
       "PG036825  PG036825.1  JP 2022536474-A/6: Methods and Compositions of...   \n",
       "PG036824  PG036824.1  JP 2022536474-A/5: Methods and Compositions of...   \n",
       "PG036823  PG036823.1  JP 2022536474-A/4: Methods and Compositions of...   \n",
       "...              ...                                                ...   \n",
       "GN354831  GN354831.1             Sequence 4595 from Patent WO2007130519   \n",
       "GN354830  GN354830.1             Sequence 4594 from Patent WO2007130519   \n",
       "GN354829  GN354829.1             Sequence 4593 from Patent WO2007130519   \n",
       "DL470173  DL470173.1  ANTISENSE ANTIVIRAL COMPOUND AND METHOD FOR TR...   \n",
       "AX348182  AX348182.1                  Sequence 52 from Patent WO0202811   \n",
       "\n",
       "         Release date       Organism name  \\\n",
       "PG036827  28-FEB-2025    Human astrovirus   \n",
       "PG036826  28-FEB-2025    Human astrovirus   \n",
       "PG036825  28-FEB-2025    Human astrovirus   \n",
       "PG036824  28-FEB-2025    Human astrovirus   \n",
       "PG036823  28-FEB-2025    Human astrovirus   \n",
       "...               ...                 ...   \n",
       "GN354831  12-MAY-2009    Mamastrovirus 13   \n",
       "GN354830  12-MAY-2009    Mamastrovirus 13   \n",
       "GN354829  12-MAY-2009    Mamastrovirus 13   \n",
       "DL470173  22-DEC-2008    Human astrovirus   \n",
       "AX348182  06-FEB-2002  Human astrovirus 1   \n",
       "\n",
       "                                              Virus Lineage  \\\n",
       "PG036827  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "PG036826  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "PG036825  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "PG036824  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "PG036823  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "...                                                     ...   \n",
       "GN354831  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "GN354830  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "GN354829  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "DL470173  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "AX348182  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "\n",
       "                        Species        Family  Length Submitters Strain  \\\n",
       "PG036827                     NA  Astroviridae    5388        NaN     NA   \n",
       "PG036826                     NA  Astroviridae    5412        NaN     NA   \n",
       "PG036825                     NA  Astroviridae    5358        NaN     NA   \n",
       "PG036824                     NA  Astroviridae    5358        NaN     NA   \n",
       "PG036823                     NA  Astroviridae    5337        NaN     NA   \n",
       "...                         ...           ...     ...        ...    ...   \n",
       "GN354831     Mamastrovirus ovis  Astroviridae      60        NaN     NA   \n",
       "GN354830     Mamastrovirus ovis  Astroviridae      60        NaN     NA   \n",
       "GN354829     Mamastrovirus ovis  Astroviridae      60        NaN     NA   \n",
       "DL470173                     NA  Astroviridae      40        NaN     NA   \n",
       "AX348182  Mamastrovirus hominis  Astroviridae      67        NaN     NA   \n",
       "\n",
       "         Isolate Geo location Country Collection date Tissue/Specimen/Source  \\\n",
       "PG036827      NA           NA      NA              NA                     NA   \n",
       "PG036826      NA           NA      NA              NA                     NA   \n",
       "PG036825      NA           NA      NA              NA                     NA   \n",
       "PG036824      NA           NA      NA              NA                     NA   \n",
       "PG036823      NA           NA      NA              NA                     NA   \n",
       "...          ...          ...     ...             ...                    ...   \n",
       "GN354831      NA           NA      NA              NA                     NA   \n",
       "GN354830      NA           NA      NA              NA                     NA   \n",
       "GN354829      NA           NA      NA              NA                     NA   \n",
       "DL470173      NA           NA      NA              NA                     NA   \n",
       "AX348182      NA           NA      NA              NA                     NA   \n",
       "\n",
       "         Host Environmental  CDS count Artificial UNVERIFIED  \n",
       "PG036827   NA            No          0        Yes         No  \n",
       "PG036826   NA            No          0        Yes         No  \n",
       "PG036825   NA            No          0        Yes         No  \n",
       "PG036824   NA            No          0        Yes         No  \n",
       "PG036823   NA            No          0        Yes         No  \n",
       "...       ...           ...        ...        ...        ...  \n",
       "GN354831   NA            No          0        Yes         No  \n",
       "GN354830   NA            No          0        Yes         No  \n",
       "GN354829   NA            No          0        Yes         No  \n",
       "DL470173   NA            No          0        Yes         No  \n",
       "AX348182   NA            No          0        Yes         No  \n",
       "\n",
       "[102 rows x 20 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[data[\"Artificial\"] == \"Yes\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "d0eab5ed-d961-4ca6-843e-11e872ef368e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "D:\\Virology\\RNA_viruses\\Astrovirus\\Astroviridae_database\\Astroviridae_15102025_orf-coords.csv\n",
      "No keywords were found in description for PX395426\n",
      "No keywords were found in description for PX395425\n",
      "No keywords were found in description for PX395424\n",
      "No keywords were found in description for PX395423\n",
      "No keywords were found in description for PX395422\n",
      "No keywords were found in description for OR951076\n",
      "No keywords were found in description for OR951065\n",
      "No keywords were found in description for OR951063\n",
      "No keywords were found in description for OR951056\n",
      "No keywords were found in description for OR951044\n",
      "No keywords were found in description for OR951043\n",
      "No keywords were found in description for OR951029\n",
      "No keywords were found in description for OR951024\n",
      "No keywords were found in description for OR951013\n",
      "No keywords were found in description for PX095754\n",
      "No keywords were found in description for PX113254\n",
      "Conflicting CDS annotations for PV999257\n",
      "Resolved for ORF1b\n",
      "Conflicting annotations in misc_feature for PV793828\n",
      "['similar to non-structural polyprotein']\n",
      "[]\n",
      "No keywords were found in description for PQ421851\n",
      "No keywords were found in description for PP211271\n",
      "No keywords were found in description for PP211222\n",
      "No keywords were found in description for PP211221\n",
      "Conflicting CDS annotations for PV524668\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV524668\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for PV400867\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for PV400866\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for PV400865\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for FJ571068\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for FJ571067\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for FJ571066\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for FJ571065\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for PV189393\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV189392\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV189391\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV189390\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV189389\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV189388\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV189387\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV189386\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV189385\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV189384\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PV189383\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for PQ541178\n",
      "Resolved for ORF1ab\n",
      "No keywords were found in description for PL092225\n",
      "No keywords were found in description for PL092217\n",
      "No keywords were found in description for PL092213\n",
      "Conflicting CDS annotations for PQ525428\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for PQ525428\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for PQ055539\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for PQ055538\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for PQ055537\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for PQ055536\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for PQ055535\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for PQ055533\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for PQ055532\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for PQ055530\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for PQ055529\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for PQ055528\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for PQ055524\n",
      "Could not resolve the conflict\n",
      "No keywords were found in description for PQ110289\n",
      "Conflicting CDS annotations for PP968017\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for OQ758048\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758047\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758046\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758045\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758044\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758043\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758042\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758041\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758040\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758039\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758038\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758037\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758036\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758035\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758034\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758033\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758032\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for OQ758031\n",
      "Resolved for ORF1b\n",
      "Conflicting annotations in misc_feature for OR871062\n",
      "['similar to hypothetical protein']\n",
      "[]\n",
      "Conflicting annotations in misc_feature for OR871061\n",
      "['similar to hypothetical protein']\n",
      "[]\n",
      "No keywords were found in description for OR871060\n",
      "No keywords were found in description for OR800315\n",
      "No keywords were found in description for OQ986647\n",
      "No keywords were found in description for OR645486\n",
      "Conflicting CDS annotations for OR626615\n",
      "Resolved for ORF1a\n",
      "No keywords were found in description for OQ802763\n",
      "No keywords were found in description for OQ802762\n",
      "No keywords were found in description for OQ802757\n",
      "No keywords were found in description for OQ802714\n",
      "No keywords were found in description for OR234613\n",
      "No keywords were found in description for OR234607\n",
      "No keywords were found in description for OQ709194\n",
      "No keywords were found in description for OQ709192\n",
      "No keywords were found in description for OQ709190\n",
      "No keywords were found in description for OQ709188\n",
      "Conflicting CDS annotations for ON400505\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for ON382522\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for ON382521\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for ON382520\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for ON382519\n",
      "Resolved for ORF1a\n",
      "Conflicting annotations in misc_feature for OQ599876\n",
      "['similar to ORF3']\n",
      "[]\n",
      "No keywords were found in description for ON932807\n",
      "Conflicting CDS annotations for OM469237\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for OM469236\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for OM469235\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for OM469234\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for OM469233\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for OM469232\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for OM469231\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for OM469230\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for OM469229\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for OP413943\n",
      "Could not resolve the conflict\n",
      "Conflicting CDS annotations for ON715001\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for ON714997\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for ON714996\n",
      "The coordinates've been extracted from another CDS\n",
      "No keywords were found in description for OP474168\n",
      "Conflicting CDS annotations for ON714999\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for ON714998\n",
      "The coordinates've been extracted from another CDS\n",
      "No keywords were found in description for OK483336\n",
      "No keywords were found in description for OK483335\n",
      "Conflicting CDS annotations for OP020144\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MW787033\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787031\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787030\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787029\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787028\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787027\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787026\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787025\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787024\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787023\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787022\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787021\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787020\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787019\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787017\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787016\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787015\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787014\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW787013\n",
      "Resolved for ORF1b\n",
      "No keywords were found in description for ON304004\n",
      "No keywords were found in description for OM953853\n",
      "Conflicting CDS annotations for MT333760\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333759\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333758\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333757\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333756\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333755\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333754\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333753\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333752\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333751\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333750\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333749\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333748\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333747\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333746\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT333745\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MZ150559\n",
      "Resolved for ORF1ab\n",
      "No keywords were found in description for OL739530\n",
      "No keywords were found in description for MW239504\n",
      "No keywords were found in description for MW239381\n",
      "No keywords were found in description for MW239289\n",
      "No keywords were found in description for MW239205\n",
      "No keywords were found in description for MW239195\n",
      "No keywords were found in description for MW239194\n",
      "Conflicting CDS annotations for MZ576222\n",
      "Resolved for ORF1a\n",
      "No keywords were found in description for MZ291967\n",
      "No keywords were found in description for MW924358\n",
      "No keywords were found in description for MW924357\n",
      "No keywords were found in description for MW924356\n",
      "Conflicting CDS annotations for MW684833\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684832\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684829\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684827\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684826\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684825\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684824\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684822\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684821\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684820\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684819\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MW684818\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MT740274\n",
      "Resolved for ORF1b\n",
      "No keywords were found in description for MW645022\n",
      "No keywords were found in description for MW645021\n",
      "Conflicting CDS annotations for MW249010\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MW592379\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MW592378\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MW592377\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for MT492440\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492439\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492438\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492437\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492436\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492435\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492434\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492433\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492432\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492431\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492430\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492429\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492428\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492427\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492426\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492425\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492424\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492423\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MT492422\n",
      "Resolved for ORF1b\n",
      "No keywords were found in description for MT896156\n",
      "No keywords were found in description for MT896155\n",
      "Conflicting CDS annotations for MT506231\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for MT506231\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123921\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123920\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123919\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123918\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123917\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123916\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123915\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123914\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123913\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123912\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123911\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123910\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123909\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123908\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123907\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123906\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123905\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123904\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123903\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123902\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123901\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123900\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123899\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123898\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123897\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123896\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MH123895\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MN725025\n",
      "Resolved for ORF2\n",
      "No keywords were found in description for MN733910\n",
      "No keywords were found in description for MN036159\n",
      "No keywords were found in description for MN036131\n",
      "No keywords were found in description for MN035893\n",
      "No keywords were found in description for MN035574\n",
      "No keywords were found in description for MN035429\n",
      "No keywords were found in description for MN035369\n",
      "No keywords were found in description for MN034849\n",
      "No keywords were found in description for MN034422\n",
      "No keywords were found in description for MN034197\n",
      "No keywords were found in description for MN033907\n",
      "No keywords were found in description for MN033591\n",
      "No keywords were found in description for MN033141\n",
      "Conflicting CDS annotations for NC_043103\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for NC_043100\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for NC_043098\n",
      "Resolved for ORF1b\n",
      "No keywords were found in description for MK387179\n",
      "No keywords were found in description for MK387178\n",
      "No keywords were found in description for MK387177\n",
      "Conflicting CDS annotations for HM484374\n",
      "Resolved for ORF1ab\n",
      "No keywords were found in description for MG846413\n",
      "Conflicting CDS annotations for NC_013443\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for MG004212\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004211\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004210\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004209\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004208\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004207\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004206\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004205\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004204\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004203\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004202\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004201\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004200\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004199\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004198\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004197\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004196\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for MG004195\n",
      "Resolved for ORF1b\n",
      "No keywords were found in description for KC411816\n",
      "MF973517\n",
      "1\n",
      "{'1A': [0, 502], '1A-strand': 1}\n",
      "Conflicting CDS annotations for GQ415660\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for KX033447\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for KY047739\n",
      "Resolved for ORF2\n",
      "Conflicting CDS annotations for KX060809\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for KX060808\n",
      "The coordinates've been extracted from another CDS\n",
      "Conflicting CDS annotations for KT963070\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KT963069\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KT757532\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KT757531\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KT757530\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KT757529\n",
      "Resolved for ORF1b\n",
      "No keywords were found in description for DQ327618\n",
      "No keywords were found in description for DQ327617\n",
      "No keywords were found in description for DQ327616\n",
      "No keywords were found in description for DQ327615\n",
      "No keywords were found in description for DQ327614\n",
      "No keywords were found in description for DQ327613\n",
      "No keywords were found in description for DQ327612\n",
      "No keywords were found in description for DQ327611\n",
      "No keywords were found in description for DQ327610\n",
      "No keywords were found in description for DQ327609\n",
      "No keywords were found in description for DQ327608\n",
      "No keywords were found in description for AY831433\n",
      "Conflicting CDS annotations for S68561\n",
      "Resolved for ORF2\n",
      "Conflicting CDS annotations for HM120878\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for HM120877\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for HM120876\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for AF290509\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for AF290508\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for AF290507\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for AF290506\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for AF290505\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for AF290504\n",
      "Resolved for ORF1a\n",
      "No keywords were found in description for AY027809\n",
      "Conflicting CDS annotations for LN879482\n",
      "Resolved for ORF1ab\n",
      "Conflicting CDS annotations for KJ496050\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ496049\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ496048\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ496047\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ496046\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ496045\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KP264970\n",
      "The coordinates've been extracted from another CDS\n",
      "No keywords were found in description for S47249\n",
      "Conflicting CDS annotations for KP404152\n",
      "Resolved for ORF2\n",
      "Conflicting CDS annotations for KP404151\n",
      "Resolved for ORF2\n",
      "Conflicting CDS annotations for KJ571485\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571484\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571483\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571482\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571481\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571480\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571479\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571478\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571477\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571476\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571475\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571474\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571473\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571472\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571471\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571470\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571469\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571468\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571467\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571466\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571465\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571464\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571463\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571462\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571461\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571460\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571459\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571458\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571457\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571456\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571455\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571454\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571453\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571452\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571451\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571450\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571449\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571448\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571447\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571446\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571445\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571444\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571443\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571442\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571441\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571440\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571439\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571438\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571437\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571436\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571435\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571434\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571433\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571432\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571431\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571430\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571429\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571428\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571427\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571426\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571425\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571424\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571423\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571422\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571421\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571420\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571419\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571418\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571417\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571416\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571415\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571414\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571413\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571412\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571411\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571410\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571409\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571408\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571407\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571406\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571405\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571404\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571403\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571402\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571401\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571400\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571399\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571398\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571397\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571396\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571395\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571394\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571393\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571392\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571391\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571390\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571389\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571388\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571387\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571386\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571385\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571384\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571383\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571382\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571381\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571380\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571379\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571378\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ571377\n",
      "Resolved for ORF1b\n",
      "No keywords were found in description for KC599256\n",
      "Conflicting CDS annotations for KJ633799\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ633798\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ633797\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ633796\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ633795\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ633794\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ633793\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ633792\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ633791\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ633790\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for KJ596489\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for JX087965\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for JF327667\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for AB518703\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for AB518702\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for AB243067\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for JX087964\n",
      "Resolved for ORF1b\n",
      "Conflicting CDS annotations for JX087963\n",
      "Resolved for ORF1b\n",
      "No keywords were found in description for HUAIMMREAB\n",
      "No keywords were found in description for HUAIMMREAA\n",
      "No keywords were found in description for HUA3TERMS\n",
      "No keywords were found in description for EF490429\n",
      "No keywords were found in description for GU222334\n",
      "No keywords were found in description for GU222333\n",
      "No keywords were found in description for GU222332\n",
      "No keywords were found in description for GU222331\n",
      "No keywords were found in description for GU222330\n",
      "No keywords were found in description for GU222348\n",
      "No keywords were found in description for GU222347\n",
      "No keywords were found in description for GU222346\n",
      "No keywords were found in description for GU222345\n",
      "No keywords were found in description for GU222344\n",
      "No keywords were found in description for GU222343\n",
      "No keywords were found in description for GU222342\n",
      "No keywords were found in description for GU222341\n",
      "No keywords were found in description for GU222340\n",
      "No keywords were found in description for GU222339\n",
      "No keywords were found in description for GU222338\n",
      "No keywords were found in description for GU222337\n",
      "No keywords were found in description for GU222336\n",
      "No keywords were found in description for GU222335\n",
      "No keywords were found in description for ATVPOLY4A\n",
      "No keywords were found in description for ATVPOLY1A\n",
      "No keywords were found in description for Z16420\n",
      "Conflicting CDS annotations for AB126674\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for AB126673\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for AB126672\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for AB126671\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for AB126670\n",
      "Resolved for ORF1a\n",
      "Conflicting CDS annotations for AB116554\n",
      "Resolved for ORF1a\n",
      "No keywords were found in description for Z11682\n",
      "Entries with no CDS annotation: 64\n",
      "Entries with no CDS with target ORF annotation: 179\n",
      "CDS with no target ORFs: 706\n",
      "CDS with no annotations: 2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "orf_coord(\"data/Astroviridae_15102025.gb\", \"ORF_names.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df8bd21a-0372-4e50-966a-9482972e3821",
   "metadata": {},
   "source": [
    "Два раза встречается ORF1ab polyprotein в PX095750"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cccee23-1e7b-472a-89d7-384677c8bd97",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
