{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "44a64e6c-37e0-4727-9c19-bf81e91d90e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO, Entrez\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "92abba5a-f22b-4b91-9c88-2a30b7ed5681",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "'''\n",
    "Downloads entries from GenBank Nucleotide by given query\n",
    "Saves entries to file in GenBank format\n",
    "Input:\n",
    "    query - str - query for search in GenBank\n",
    "Output:\n",
    "    outfile - str - shortname of output file with sequences in genbank format\n",
    "'''\n",
    "def fetch_seq_from_Nucleotide(query, outfile):\n",
    "\n",
    "    Entrez.email = \"A.N.Other@example.com\"\n",
    "    # list with ids obtained by query\n",
    "    record = Entrez.read(Entrez.esearch(db=\"nucleotide\", term=query, idtype=\"acc\", RetMax=1000000))\n",
    "    id_list = record['IdList']\n",
    "    print(\"Query to GenBank Nucleotide database:\\\"{}\\\"\".format(query))\n",
    "    print(\"Number of records found: {}\".format(record[\"Count\"]))\n",
    "    print(\"{} records will be downloaded\".format(len(id_list)))\n",
    "    # output file with sequences will be saved in current working dir\n",
    "    f_name = os.path.abspath(outfile)\n",
    "\n",
    "    # output file with sequences in fasta format\n",
    "    file_out = open(f_name, 'w')\n",
    "    \n",
    "    # Saving sequences \n",
    "    \n",
    "    i = 0\n",
    "    # fetch each sequence from GenBank by its ID and writes to ouput file\n",
    "    for id in id_list:\n",
    "        i+=1\n",
    "        if i % 500 ==0:\n",
    "            print(\"Downloaded {} sequences\".format(i))\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=id, rettype=\"genbank\", retmode=\"text\")\n",
    "        for line in handle:\n",
    "            file_out.write(line)\n",
    "        handle.close()\n",
    "    file_out.close()\n",
    "    print('Finished')\n",
    "    return(f_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a7bb4ed5-1832-4676-8515-a9339e4e1f5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_metadata_from_gb(input_file):\n",
    "    '''\n",
    "    For each entry in GenBank file retrieves the following data:\n",
    "\n",
    "    # Identifiers\n",
    "    Accession\n",
    "    GenBank title (DEFINITION)\n",
    "\n",
    "    # Classification\n",
    "    Organism name\n",
    "    Species\n",
    "    Isolate\n",
    "    Strain\n",
    "    Family\n",
    "    Lineage\n",
    "\n",
    "    #Sequence quality\n",
    "    Length\n",
    "    Strand\n",
    "\n",
    "    Num of annotated CDS\n",
    "\n",
    "    ORF1a\n",
    "    ORF1b\n",
    "    ORF2\n",
    "    coords\n",
    "    \n",
    "    #Sources\n",
    "    Country\n",
    "    Tissue/Specimen/Source\n",
    "    Collection date\n",
    "    Release date\n",
    "    Submitters\n",
    "    \n",
    "    '''\n",
    "\n",
    "    '''\n",
    "     dictionary\n",
    "     entries_data[GenBank Accession] = {}\n",
    "     entries_data[GenBank Accession].keys() = [\n",
    "                                                'Version',\n",
    "                                                'GenBank title'\n",
    "                                                'Organism name',\n",
    "                                                'Species',\n",
    "                                                'Family',\n",
    "                                                'Virus Lineage',\n",
    "                                                'Length',\n",
    "                                                'Isolate',\n",
    "                                                'Strain',\n",
    "                                                'Strand',\n",
    "                                                'Geo location'\n",
    "                                                'Country',\n",
    "                                                'Tissue/Specimen/Source',\n",
    "                                                'Host',\n",
    "                                                'Collection date',\n",
    "                                                'Release date',\n",
    "                                                'Submitters',\n",
    "                                                'Sequencing Technology'\n",
    "                                                'CDS number'\n",
    "                                                ]\n",
    "    '''\n",
    "    entries_data = {}\n",
    "\n",
    "    source_qualifiers = [\"strain\", \"isolate\", \"geo_loc_name\", \"country\", \"collection_date\", \"isolation_source\", \"host\"]\n",
    "    source_qualifiers_columns = [\"Strain\", \"Isolate\", \"Geo location\", \"Country\", \"Collection date\", \"Tissue/Specimen/Source\", \"Host\"]\n",
    "\n",
    "    entries = SeqIO.parse(os.path.abspath(input_file), \"genbank\")\n",
    "\n",
    "    for entry in entries:\n",
    "        entries_data[entry.name] = {}\n",
    "        entries_data[entry.name]['Version'] = entry.id\n",
    "        entries_data[entry.name]['GenBank title'] = entry.description\n",
    "\n",
    "        entries_data[entry.name]['Release date'] = entry.annotations['date']\n",
    "        entries_data[entry.name]['Organism name'] = entry.annotations['organism']\n",
    "        \n",
    "        entries_data[entry.name]['Virus Lineage'] = ';'.join(entry.annotations['taxonomy'])\n",
    "        num_taxa = len(entry.annotations['taxonomy'])\n",
    "        if num_taxa == 9:\n",
    "            entries_data[entry.name]['Species'] = entry.annotations['taxonomy'][-1]\n",
    "        else:\n",
    "             entries_data[entry.name]['Species'] = 'NA'\n",
    "        entries_data[entry.name]['Family'] = entry.annotations['taxonomy'][6]\n",
    "        \n",
    "        entries_data[entry.name]['Length'] = len(entry)\n",
    "\n",
    "        for reference in entry.annotations['references']:\n",
    "            if reference.title == 'Direct Submission':\n",
    "                entries_data[entry.name]['Submitters'] = reference.authors\n",
    "\n",
    "        count_cds = 0\n",
    "        for feature in entry.features:\n",
    "            if feature.type == 'source':\n",
    "                for qualif, colname in zip(source_qualifiers, source_qualifiers_columns):\n",
    "                    if qualif in feature.qualifiers:\n",
    "                        entries_data[entry.name][colname] = feature.qualifiers[qualif][0]\n",
    "                    else:\n",
    "                        entries_data[entry.name][colname] = 'NA'\n",
    "                if 'environmental_sample' in feature.qualifiers:\n",
    "                    entries_data[entry.name]['Environmental'] = 'Yes'\n",
    "                else:\n",
    "                    entries_data[entry.name]['Environmental'] = 'No'\n",
    "                    \n",
    "            if feature.type == 'CDS':\n",
    "                \n",
    "                count_cds +=1\n",
    "                \n",
    "        entries_data[entry.name]['CDS count'] = count_cds\n",
    "\n",
    "    meta_dataframe = pd.DataFrame.from_dict(entries_data, orient='index')\n",
    "    \n",
    "    meta_dataframe.to_csv(os.path.splitext(os.path.abspath(input_file))[0] + '.csv')\n",
    "    #country_map - file with abbreviations of countries\n",
    "    return meta_dataframe\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "87cfd169-5ef8-4943-9047-98f218a3b9fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_feature(feature, feature_map):\n",
    "    '''\n",
    "    feature_map - dictionary, e.g. feature_map['Italia']='ITA'\n",
    "    feature - possible key from feature_map\n",
    "    return value if key is in feature_map\n",
    "    '''\n",
    "    for k, v in feature_map.items():\n",
    "        if feature.lower() == k.lower():\n",
    "            return v\n",
    "    return feature\n",
    "\n",
    "\n",
    "def orf_coord(input_file, orf_map):\n",
    "    '''\n",
    "    Retrieves the coordinates of ORFs\n",
    "    \n",
    "    Input:\n",
    "        input_file - file with nucleotide sequences in genbank-format\n",
    "        orf_map - csv file annotation of orfs and their codes\n",
    "    Output:\n",
    "        coord_file - file with coordinates\n",
    "    '''\n",
    "\n",
    "    # dictionary with possiple names of astrovirus ORFs indicated in note, gene and product qualifiers\n",
    "    # orf_dict[\"ORF1a\"] = \"1A\"\n",
    "    orf_dict = read_csv(orf_map)\n",
    "\n",
    "    # ORFs which coordinates this script will extract\n",
    "    orf_types_final = ['1A', '1B', '2']\n",
    "    # ORFs that can be met\n",
    "    orf_types = ['1A', '1B', '1AB', '2']\n",
    "\n",
    "    # name of output file with ORF coordinates\n",
    "    out_file_name_temp = os.path.splitext(os.path.abspath(input_file))[0]\n",
    "    out_file_name = out_file_name_temp  + '_orf-coords.csv'\n",
    "    print(out_file_name)\n",
    "\n",
    "    # ENTRIES with known errors in annotation\n",
    "    exceptions_ORF2 = [\"MN725025\", \"KY047739\", \"S68561\", \"KP404151\",\"KP404152\"] # the last CDS in conflict ORF1b ORF2 is ORF2\n",
    "    #AB518702 AB518703 is actually RdRp\n",
    "    exceptions_1a = []\n",
    "\n",
    "    # dictionaruy with ORF coordinates\n",
    "    dict_coord = {}\n",
    "\n",
    "    with open(input_file) as handle:\n",
    "        records = list(SeqIO.parse(handle, 'gb'))\n",
    "\n",
    "        # entries with no annotations\n",
    "        no_annot = 0\n",
    "        # entries with no annotations of target CDS\n",
    "        no_annot_target = 0\n",
    "        # CDS with no annotations of target ORFs\n",
    "        notarget_annot_CDS = 0\n",
    "        # CDS with no annotations\n",
    "        no_annot_CDS = 0\n",
    "\n",
    "        # Entries with CDS that has annotation \"nonstructural protein\"\n",
    "        entries_onlynonstrprotein = []\n",
    "        entries_onlynonstrprotein_short = []\n",
    "        entries_conflicting_annotations = []\n",
    "        entries_noannot = []\n",
    "        entries_notarget_annot = []\n",
    "        \n",
    "        for rec in records:\n",
    "\n",
    "            if \"Patent\" in rec.description or \"Modified Microbial Nucleic Acid\" in rec.description: \n",
    "                continue\n",
    "            \n",
    "            dict_coord[rec.name] = {}\n",
    "            # iterate over record features\n",
    "\n",
    "            # number of CDS\n",
    "            cds_count = 0\n",
    "            # number of annotated CDS with target ORFs\n",
    "            cds_annot_count = 0\n",
    "            for feature in rec.features:\n",
    "                \n",
    "                if feature.type == 'CDS':\n",
    "                    cds_count += 1\n",
    "                    \n",
    "                    # Check codon start\n",
    "                    if 'codon_start' in feature.qualifiers.keys():\n",
    "                        cod_start = int(feature.qualifiers['codon_start'][0]) - 1\n",
    "                        if cod_start<0:\n",
    "                            print('Codon start < 0, something is wrong')\n",
    "                    else:\n",
    "                        cod_start = 0\n",
    "\n",
    "                    # check all annotations in gene, product, note qualifiers\n",
    "                    CDS_raw_annotations = []\n",
    "                    for el in ('product', 'gene', 'note'):\n",
    "                        CDS_raw_annotations.extend(feature.qualifiers.get(el, []))\n",
    "                    \n",
    "                    if len(CDS_raw_annotations) != 0:\n",
    "                        # now we will leave annotations of target ORFs\n",
    "                        CDS_annotations = list(map(lambda x: map_feature(x, orf_dict), CDS_raw_annotations))\n",
    "                        CDS_annotations = list(filter(lambda x: x in orf_types, CDS_annotations))\n",
    "                        if len(CDS_annotations) == 0:\n",
    "                            #print(\"No annotations of target ORFs of CDS in {}\".format(rec.name))\n",
    "                            #print(CDS_raw_annotations)\n",
    "                            #print(rec.description)\n",
    "                            notarget_annot_CDS +=1\n",
    "\n",
    "                            if \"nonstructural protein\" in CDS_raw_annotations:\n",
    "                                #print(\"CDS of {} has the only annotation \\'nonstructural protein\\' \".format(rec.name))\n",
    "                                if len(rec) > 300:\n",
    "                                    entries_onlynonstrprotein.append(rec.name + ':' + rec.description + '\\n')\n",
    "                                else:\n",
    "                                    entries_onlynonstrprotein_short.append(rec.name + ':' + rec.description + '\\n')\n",
    "                        else:\n",
    "                            #print(\"Annotations were successfully found\")\n",
    "                            #print(CDS_annotations)\n",
    "\n",
    "                            # check for consistency of CDS annotations\n",
    "                            uniq_annotations = list(set(CDS_annotations))\n",
    "                            if len(uniq_annotations) == 1:\n",
    "                                CDS_product = uniq_annotations[0]\n",
    "                                # CDS is a target ORF, the coordinates are not joined\n",
    "                                if CDS_product in orf_types_final:\n",
    "                                    dict_coord[rec.name][CDS_product] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                    dict_coord[rec.name][CDS_product + '-chain'] = feature.location.strand\n",
    "                                    cds_annot_count +=1\n",
    "                                elif CDS_product == \"1AB\":\n",
    "\n",
    "                                    # check whether the coordinates are joined\n",
    "                                    if len(feature.location.parts) == 2:\n",
    "                                        dict_coord[rec.name]['1A'] = [int(feature.location.parts[0].start) + cod_start, int(feature.location.parts[0].end)]\n",
    "                                        dict_coord[rec.name]['1A-chain'] = feature.location.parts[0].strand\n",
    "                                        dict_coord[rec.name]['1B'] = [int(feature.location.parts[1].start), int(feature.location.parts[1].end)]\n",
    "                                        dict_coord[rec.name]['1B-chain'] = feature.location.parts[1].strand\n",
    "                                        cds_annot_count +=1\n",
    "                                    elif len(feature.location.parts) == 1:\n",
    "                                        dict_coord[rec.name]['1B'] = [int(feature.location.parts[0].start) + cod_start, int(feature.location.parts[0].end)]\n",
    "                                        dict_coord[rec.name]['1B-chain'] = feature.location.parts[0].strand\n",
    "                                        cds_annot_count +=1\n",
    "                                    else:\n",
    "                                        print(\"strange locations in {}\".format(rec.name))\n",
    "                                        print(feature.location)\n",
    "                            else:\n",
    "                                print(\"Conflicting CDS annotations for {}\".format(rec.name))\n",
    "                                #print(CDS_annotations)\n",
    "                                #print(uniq_annotations)\n",
    "\n",
    "                                present_1A = ('1A' in uniq_annotations)\n",
    "                                present_1B = ('1B' in uniq_annotations)\n",
    "                                present_1A_1B = present_1A and present_1B\n",
    "                                present_1AB = ('1AB' in uniq_annotations)\n",
    "                                present_1A_1AB = present_1A and present_1AB\n",
    "                                present_1B_1AB = present_1B and present_1AB\n",
    "                                present_2 = ('2' in uniq_annotations)\n",
    "                                present_1AB_2 = (present_1A and present_2) or (present_1B and present_2) or (present_1AB and present_2) \n",
    "                                \n",
    "                                \n",
    "                                if present_1AB_2:\n",
    "                                    if rec.name in exceptions_ORF2:\n",
    "                                        dict_coord[rec.name]['2'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                        dict_coord[rec.name]['2-chain'] = feature.location.strand\n",
    "                                        cds_annot_count +=1\n",
    "                                        print(\"Resolved for ORF2\")\n",
    "                                    else:\n",
    "                                        print(CDS_annotations)\n",
    "                                        print(uniq_annotations)\n",
    "                                        print(\"Could not resolve the conflict\")\n",
    "                                    \n",
    "                                        entries_conflicting_annotations.append(rec.name + ': ORF1ab + ORF2' + ':' + rec.description + '\\n')\n",
    "                                # If ORF1a and ORF1b are present, we choose ORF1b\n",
    "                                elif present_1A_1B:\n",
    "                                    if rec.name in exceptions_1a:\n",
    "                                        dict_coord[rec.name]['1A'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                        dict_coord[rec.name]['1A-chain'] = feature.location.strand\n",
    "                                        cds_annot_count +=1\n",
    "                                        print(\"Resolved for ORF1a\")\n",
    "                                    else:\n",
    "                                        dict_coord[rec.name]['1B'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                        dict_coord[rec.name]['1B-chain'] = feature.location.strand\n",
    "                                        cds_annot_count +=1\n",
    "                                        entries_conflicting_annotations.append(rec.name + ': ORF1a + ORF1b' + ':' + rec.description + '\\n')\n",
    "                                        print(\"Resolved for ORF1b\")\n",
    "                                # ORF1ab with joined coordinates\n",
    "                                elif len(feature.location.parts) == 2:\n",
    "                                    dict_coord[rec.name]['1A'] = [int(feature.location.parts[0].start) + cod_start, int(feature.location.parts[0].end)]\n",
    "                                    dict_coord[rec.name]['1A-chain'] = feature.location.parts[0].strand\n",
    "                                    dict_coord[rec.name]['1B'] = [int(feature.location.parts[1].start), int(feature.location.parts[1].end)]\n",
    "                                    dict_coord[rec.name]['1B-chain'] = feature.location.parts[1].strand\n",
    "                                    cds_annot_count +=1\n",
    "                                    print(\"Resolved for ORF1ab\")\n",
    "                                elif present_1A_1AB and ('1A' not in dict_coord[rec.name].keys()):\n",
    "                                    dict_coord[rec.name]['1A'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                    dict_coord[rec.name]['1A-chain'] = feature.location.strand\n",
    "                                    cds_annot_count +=1\n",
    "                                    print(\"Resolved for ORF1a\")\n",
    "                                elif present_1B_1AB and ('1B' not in dict_coord[rec.name].keys()):\n",
    "                                    dict_coord[rec.name]['1B'] = [int(feature.location.start) + cod_start, int(feature.location.end)]\n",
    "                                    dict_coord[rec.name]['1B-chain'] = feature.location.strand\n",
    "                                    cds_annot_count +=1\n",
    "                                    print(\"Resolved for ORF1b\")\n",
    "                                elif '1B' in dict_coord[rec.name].keys() or '1A' in dict_coord[rec.name].keys():\n",
    "                                    print(\"The coordinates've been extracted from another CDS\")\n",
    "\n",
    "                                \n",
    "                                    \n",
    "                        \n",
    "                    else:\n",
    "                        #print(\"No annotations of CDS in {}\".format(rec.name))\n",
    "                        #print(rec.description)\n",
    "                        #print(feature)\n",
    "                        no_annot_CDS +=1\n",
    "            \n",
    "            if cds_count == 0:\n",
    "                #print(\"{} has no CDS in annotation\".format(rec.name))\n",
    "                #print(rec.description)\n",
    "                no_annot +=1\n",
    "                entries_noannot.append(rec.name + ':' + rec.description + '\\n')\n",
    "            if cds_annot_count == 0:\n",
    "                #print(\"{} has no CDS of target ORFs\".format(rec.name))\n",
    "                #no_annot_target +=1\n",
    "                #print(rec.description)\n",
    "                entries_notarget_annot.append(rec.name + ':' + rec.description + '\\n')\n",
    "\n",
    "\n",
    "\n",
    "        # Write extracted coordinates to file\n",
    "        with open(out_file_name, 'w') as out_file:\n",
    "            line = \"GBAC\"\n",
    "            for ORF in orf_types_final:\n",
    "                line = line + ',' + ORF + ',' + ORF + '-chain'\n",
    "            out_file.write(line + '\\n')\n",
    "            for rec_id, values in dict_coord.items():\n",
    "                s = rec_id\n",
    "                for orf in orf_types_final:\n",
    "                    if orf in values:\n",
    "                        s += f\",{values[orf][0]}-{values[orf][1]},{values[orf+'-chain']}\"\n",
    "                    else:\n",
    "                        s += \",NA-NA,NA\"\n",
    "                s += \"\\n\"\n",
    "                out_file.write(s)\n",
    "        out_file.close()\n",
    "\n",
    "\n",
    "        print(\"Entries with no CDS annotation: {}\".format(no_annot))\n",
    "        print(\"Entries with no CDS with target ORF annotation: {}\".format(no_annot_target))\n",
    "        print(\"CDS with no target ORFs: {}\".format(notarget_annot_CDS))\n",
    "        print(\"CDS with no annotations: {}\".format(no_annot_CDS))\n",
    "\n",
    "        with open(out_file_name_temp + '_noannotCDS.txt', 'w') as file:\n",
    "            file.writelines(entries_noannot)\n",
    "\n",
    "        with open(out_file_name_temp + '_noannot_targetCDS.txt', 'w') as file:\n",
    "            file.writelines(entries_notarget_annot)\n",
    "            \n",
    "        with open(out_file_name_temp + '_noannot_targetCDS.txt', 'w') as file:\n",
    "            file.writelines(entries_notarget_annot)\n",
    "\n",
    "        with open(out_file_name_temp + '_conflictingannot.txt', 'w') as file:\n",
    "            file.writelines(entries_conflicting_annotations)\n",
    "\n",
    "        with open(out_file_name_temp + '_onlyNSP_long.txt', 'w') as file:\n",
    "            file.writelines(entries_onlynonstrprotein)\n",
    "            \n",
    "        with open(out_file_name_temp + '_onlyNSP_short.txt', 'w') as file:\n",
    "            file.writelines(entries_onlynonstrprotein_short)\n",
    "\n",
    "        \n",
    "        return 0\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4bf1e178-edbb-46b3-99db-0a4d954003ee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query to GenBank Nucleotide database:\"\"txid39733\"[Organism]\"\n",
      "Number of records found: 16012\n",
      "16012 records will be downloaded\n",
      "Downloaded 500 sequences\n",
      "Downloaded 1000 sequences\n",
      "Downloaded 1500 sequences\n",
      "Downloaded 2000 sequences\n",
      "Downloaded 2500 sequences\n",
      "Downloaded 3000 sequences\n",
      "Downloaded 3500 sequences\n",
      "Downloaded 4000 sequences\n",
      "Downloaded 4500 sequences\n",
      "Downloaded 5000 sequences\n",
      "Downloaded 5500 sequences\n",
      "Downloaded 6000 sequences\n",
      "Downloaded 6500 sequences\n",
      "Downloaded 7000 sequences\n",
      "Downloaded 7500 sequences\n",
      "Downloaded 8000 sequences\n",
      "Downloaded 8500 sequences\n",
      "Downloaded 9000 sequences\n",
      "Downloaded 9500 sequences\n",
      "Downloaded 10000 sequences\n",
      "Downloaded 10500 sequences\n",
      "Downloaded 11000 sequences\n",
      "Downloaded 11500 sequences\n",
      "Downloaded 12000 sequences\n",
      "Downloaded 12500 sequences\n",
      "Downloaded 13000 sequences\n",
      "Downloaded 13500 sequences\n",
      "Downloaded 14000 sequences\n",
      "Downloaded 14500 sequences\n",
      "Downloaded 15000 sequences\n",
      "Downloaded 15500 sequences\n",
      "Downloaded 16000 sequences\n",
      "Finished\n"
     ]
    }
   ],
   "source": [
    "file_gb = fetch_seq_from_Nucleotide('\"txid39733\"[Organism]' , \"Astroviridae_15102025.gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e8272523-aba2-49da-a300-12cb03eacb19",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = \"Astroviridae_15102025.gb\"\n",
    "data = fetch_metadata_from_gb(input_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "004f9fb5-e731-4106-8415-87dca8049f4c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Version</th>\n",
       "      <th>GenBank title</th>\n",
       "      <th>Release date</th>\n",
       "      <th>Organism name</th>\n",
       "      <th>Virus Lineage</th>\n",
       "      <th>Species</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Submitters</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Isolate</th>\n",
       "      <th>Geo location</th>\n",
       "      <th>Country</th>\n",
       "      <th>Collection date</th>\n",
       "      <th>Tissue/Specimen/Source</th>\n",
       "      <th>Host</th>\n",
       "      <th>Environmental</th>\n",
       "      <th>CDS count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PV418556</th>\n",
       "      <td>PV418556.1</td>\n",
       "      <td>Canine astrovirus isolate CaAstV/Dog/AUS/hunte...</td>\n",
       "      <td>08-OCT-2025</td>\n",
       "      <td>Canine astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>6444</td>\n",
       "      <td>Mifsud,J.C.O., Harvey,E., Van Brussel,K., Olss...</td>\n",
       "      <td>NA</td>\n",
       "      <td>CaAstV/Dog/AUS/huntervalley_7/2022</td>\n",
       "      <td>Australia: Hunter Valley</td>\n",
       "      <td>NA</td>\n",
       "      <td>Aug-2022</td>\n",
       "      <td>feces</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>No</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PV418555</th>\n",
       "      <td>PV418555.1</td>\n",
       "      <td>Canine astrovirus isolate CaAstV/Dog/AUS/hunte...</td>\n",
       "      <td>08-OCT-2025</td>\n",
       "      <td>Canine astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>6466</td>\n",
       "      <td>Mifsud,J.C.O., Harvey,E., Van Brussel,K., Olss...</td>\n",
       "      <td>NA</td>\n",
       "      <td>CaAstV/Dog/AUS/huntervalley_6/2022</td>\n",
       "      <td>Australia: Hunter Valley</td>\n",
       "      <td>NA</td>\n",
       "      <td>Aug-2022</td>\n",
       "      <td>feces</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>No</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PV418548</th>\n",
       "      <td>PV418548.1</td>\n",
       "      <td>Canine astrovirus isolate CaAstV/Dog/AUS/b_hop...</td>\n",
       "      <td>08-OCT-2025</td>\n",
       "      <td>Canine astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>2447</td>\n",
       "      <td>Mifsud,J.C.O., Harvey,E., Van Brussel,K., Olss...</td>\n",
       "      <td>NA</td>\n",
       "      <td>CaAstV/Dog/AUS/b_hopevale_4/2021</td>\n",
       "      <td>Australia: Hope Vale</td>\n",
       "      <td>NA</td>\n",
       "      <td>Dec-2021</td>\n",
       "      <td>rectal swab</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PV418547</th>\n",
       "      <td>PV418547.1</td>\n",
       "      <td>Canine astrovirus isolate CaAstV/Dog/AUS/a_hop...</td>\n",
       "      <td>08-OCT-2025</td>\n",
       "      <td>Canine astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>2416</td>\n",
       "      <td>Mifsud,J.C.O., Harvey,E., Van Brussel,K., Olss...</td>\n",
       "      <td>NA</td>\n",
       "      <td>CaAstV/Dog/AUS/a_hopevale_4/2021</td>\n",
       "      <td>Australia: Hope Vale</td>\n",
       "      <td>NA</td>\n",
       "      <td>Dec-2021</td>\n",
       "      <td>rectal swab</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PV418557</th>\n",
       "      <td>PV418557.1</td>\n",
       "      <td>Canine astrovirus isolate CaAstV/Dog/AUS/perth...</td>\n",
       "      <td>08-OCT-2025</td>\n",
       "      <td>Canine astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>6478</td>\n",
       "      <td>Mifsud,J.C.O., Harvey,E., Van Brussel,K., Olss...</td>\n",
       "      <td>NA</td>\n",
       "      <td>CaAstV/Dog/AUS/perth_6/2022</td>\n",
       "      <td>Australia: Perth</td>\n",
       "      <td>NA</td>\n",
       "      <td>Mar-2022</td>\n",
       "      <td>feces</td>\n",
       "      <td>Canis lupus familiaris</td>\n",
       "      <td>No</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y08628</th>\n",
       "      <td>Y08628.1</td>\n",
       "      <td>Human astrovirus type 2 gene encoding capsid p...</td>\n",
       "      <td>26-JUL-2016</td>\n",
       "      <td>Human astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>302</td>\n",
       "      <td>Monceyron,C.</td>\n",
       "      <td>NA</td>\n",
       "      <td>sample 6190</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y08627</th>\n",
       "      <td>Y08627.1</td>\n",
       "      <td>Human astrovirus type 1 gene encoding capsid p...</td>\n",
       "      <td>26-JUL-2016</td>\n",
       "      <td>Human astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>300</td>\n",
       "      <td>Monceyron,C.</td>\n",
       "      <td>NA</td>\n",
       "      <td>sample 7783</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z66541</th>\n",
       "      <td>Z66541.1</td>\n",
       "      <td>Human astrovirus type 8 orf2 gene for capsid p...</td>\n",
       "      <td>30-OCT-1995</td>\n",
       "      <td>Human astrovirus 8</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>Mamastrovirus hominis</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>2434</td>\n",
       "      <td>Carter,M.J.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z46658</th>\n",
       "      <td>Z46658.1</td>\n",
       "      <td>Human astrovirus type 6 gene for capsid polypr...</td>\n",
       "      <td>14-NOV-1994</td>\n",
       "      <td>Human astrovirus 6</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>Mamastrovirus hominis</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>2418</td>\n",
       "      <td>Willcocks,M.M.</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HUANSSPS</th>\n",
       "      <td>L13745.1</td>\n",
       "      <td>Human astrovirus serotype 2, complete sequence...</td>\n",
       "      <td>08-MAR-1994</td>\n",
       "      <td>Human astrovirus</td>\n",
       "      <td>Viruses;Riboviria;Orthornavirae;Pisuviricota;S...</td>\n",
       "      <td>NA</td>\n",
       "      <td>Astroviridae</td>\n",
       "      <td>6828</td>\n",
       "      <td>Jiang,B.</td>\n",
       "      <td>Oxford</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>No</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16012 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Version                                      GenBank title  \\\n",
       "PV418556  PV418556.1  Canine astrovirus isolate CaAstV/Dog/AUS/hunte...   \n",
       "PV418555  PV418555.1  Canine astrovirus isolate CaAstV/Dog/AUS/hunte...   \n",
       "PV418548  PV418548.1  Canine astrovirus isolate CaAstV/Dog/AUS/b_hop...   \n",
       "PV418547  PV418547.1  Canine astrovirus isolate CaAstV/Dog/AUS/a_hop...   \n",
       "PV418557  PV418557.1  Canine astrovirus isolate CaAstV/Dog/AUS/perth...   \n",
       "...              ...                                                ...   \n",
       "Y08628      Y08628.1  Human astrovirus type 2 gene encoding capsid p...   \n",
       "Y08627      Y08627.1  Human astrovirus type 1 gene encoding capsid p...   \n",
       "Z66541      Z66541.1  Human astrovirus type 8 orf2 gene for capsid p...   \n",
       "Z46658      Z46658.1  Human astrovirus type 6 gene for capsid polypr...   \n",
       "HUANSSPS    L13745.1  Human astrovirus serotype 2, complete sequence...   \n",
       "\n",
       "         Release date       Organism name  \\\n",
       "PV418556  08-OCT-2025   Canine astrovirus   \n",
       "PV418555  08-OCT-2025   Canine astrovirus   \n",
       "PV418548  08-OCT-2025   Canine astrovirus   \n",
       "PV418547  08-OCT-2025   Canine astrovirus   \n",
       "PV418557  08-OCT-2025   Canine astrovirus   \n",
       "...               ...                 ...   \n",
       "Y08628    26-JUL-2016    Human astrovirus   \n",
       "Y08627    26-JUL-2016    Human astrovirus   \n",
       "Z66541    30-OCT-1995  Human astrovirus 8   \n",
       "Z46658    14-NOV-1994  Human astrovirus 6   \n",
       "HUANSSPS  08-MAR-1994    Human astrovirus   \n",
       "\n",
       "                                              Virus Lineage  \\\n",
       "PV418556  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "PV418555  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "PV418548  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "PV418547  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "PV418557  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "...                                                     ...   \n",
       "Y08628    Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "Y08627    Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "Z66541    Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "Z46658    Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "HUANSSPS  Viruses;Riboviria;Orthornavirae;Pisuviricota;S...   \n",
       "\n",
       "                        Species        Family  Length  \\\n",
       "PV418556                     NA  Astroviridae    6444   \n",
       "PV418555                     NA  Astroviridae    6466   \n",
       "PV418548                     NA  Astroviridae    2447   \n",
       "PV418547                     NA  Astroviridae    2416   \n",
       "PV418557                     NA  Astroviridae    6478   \n",
       "...                         ...           ...     ...   \n",
       "Y08628                       NA  Astroviridae     302   \n",
       "Y08627                       NA  Astroviridae     300   \n",
       "Z66541    Mamastrovirus hominis  Astroviridae    2434   \n",
       "Z46658    Mamastrovirus hominis  Astroviridae    2418   \n",
       "HUANSSPS                     NA  Astroviridae    6828   \n",
       "\n",
       "                                                 Submitters  Strain  \\\n",
       "PV418556  Mifsud,J.C.O., Harvey,E., Van Brussel,K., Olss...      NA   \n",
       "PV418555  Mifsud,J.C.O., Harvey,E., Van Brussel,K., Olss...      NA   \n",
       "PV418548  Mifsud,J.C.O., Harvey,E., Van Brussel,K., Olss...      NA   \n",
       "PV418547  Mifsud,J.C.O., Harvey,E., Van Brussel,K., Olss...      NA   \n",
       "PV418557  Mifsud,J.C.O., Harvey,E., Van Brussel,K., Olss...      NA   \n",
       "...                                                     ...     ...   \n",
       "Y08628                                         Monceyron,C.      NA   \n",
       "Y08627                                         Monceyron,C.      NA   \n",
       "Z66541                                          Carter,M.J.      NA   \n",
       "Z46658                                       Willcocks,M.M.      NA   \n",
       "HUANSSPS                                           Jiang,B.  Oxford   \n",
       "\n",
       "                                     Isolate              Geo location  \\\n",
       "PV418556  CaAstV/Dog/AUS/huntervalley_7/2022  Australia: Hunter Valley   \n",
       "PV418555  CaAstV/Dog/AUS/huntervalley_6/2022  Australia: Hunter Valley   \n",
       "PV418548    CaAstV/Dog/AUS/b_hopevale_4/2021      Australia: Hope Vale   \n",
       "PV418547    CaAstV/Dog/AUS/a_hopevale_4/2021      Australia: Hope Vale   \n",
       "PV418557         CaAstV/Dog/AUS/perth_6/2022          Australia: Perth   \n",
       "...                                      ...                       ...   \n",
       "Y08628                           sample 6190                        NA   \n",
       "Y08627                           sample 7783                        NA   \n",
       "Z66541                                    NA                        NA   \n",
       "Z46658                                    NA                        NA   \n",
       "HUANSSPS                                  NA                        NA   \n",
       "\n",
       "         Country Collection date Tissue/Specimen/Source  \\\n",
       "PV418556      NA        Aug-2022                  feces   \n",
       "PV418555      NA        Aug-2022                  feces   \n",
       "PV418548      NA        Dec-2021            rectal swab   \n",
       "PV418547      NA        Dec-2021            rectal swab   \n",
       "PV418557      NA        Mar-2022                  feces   \n",
       "...          ...             ...                    ...   \n",
       "Y08628        NA              NA                     NA   \n",
       "Y08627        NA              NA                     NA   \n",
       "Z66541        NA              NA                     NA   \n",
       "Z46658        NA              NA                     NA   \n",
       "HUANSSPS      NA              NA                     NA   \n",
       "\n",
       "                            Host Environmental  CDS count  \n",
       "PV418556  Canis lupus familiaris            No          3  \n",
       "PV418555  Canis lupus familiaris            No          3  \n",
       "PV418548  Canis lupus familiaris            No          1  \n",
       "PV418547  Canis lupus familiaris            No          1  \n",
       "PV418557  Canis lupus familiaris            No          3  \n",
       "...                          ...           ...        ...  \n",
       "Y08628                        NA            No          1  \n",
       "Y08627                        NA            No          1  \n",
       "Z66541                        NA            No          1  \n",
       "Z46658                        NA            No          1  \n",
       "HUANSSPS                      NA            No          0  \n",
       "\n",
       "[16012 rows x 18 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0eab5ed-d961-4ca6-843e-11e872ef368e",
   "metadata": {},
   "outputs": [],
   "source": [
    "orf_coord(\"Astroviridae_15102025_template.gb\", \"ORF_names.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fd2e5809-d4be-4196-ad07-6aca075cf232",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "D:\\Virology\\RNA_viruses\\Astroviridae_database\\Astroviridae.gb\n"
     ]
    }
   ],
   "source": [
    "outfile = \"Astroviridae_15102025.gb\"\n",
    "outfile = \"D:\\Virology\\RNA_viruses\\Astroviridae_database\\Astroviridae.gb\"\n",
    "print(os.path.abspath(outfile))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1697f50b-2d3f-410d-a8cc-2705b51fd661",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = \"Astroviridae_15102025_template.gb\"\n",
    "data = fetch_metadata_from_gb(input_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
